### PR TITLE
WIP 40 xss amelioration

### LINKED ISSUE #40: Potential XSS Vulnerability
In PostMessageBridge, we do not test that the origin prop. matches what is expected. Theoretically an attacker could send arbitrary JS to execute in the listener window.

https://github.com/screencloud/app-core/blob/fb1d8766935ec0d30281fbdaa6dace5a99b382e1/src/PostMessageBridge.ts#L132-L136

Further, `this.target` is defined as:

https://github.com/screencloud/app-core/blob/fb1d8766935ec0d30281fbdaa6dace5a99b382e1/src/PostMessageBridge.ts#L47

Suggested resolution:
1. tighten up the initializer for targetWindow
2. 1. Use something like this to validate `event.origin` in addition to checking `event.source`.  (These domains may not be totally correct).
```
function validOrigin(origin) {
    const validPostMessageDomains = [
        "screen\.cloud",
        "screencloudapp\.com",
        "screencloud\.com",
        "screencloudapps\.com",
    ];
    const re = RegExp(`^https://(${validPostMessageDomains.join("|")})$`, "i")
    // Just for readability, written as an array and joined rather than one long regex
    // Be careful to escape every "."

    return re.test(origin)
}
```
@wmathes @michael-martin

### PR DESCRIPTION
**Pull request by Macavirus**
# Looking for feedback :small_airplane: :small_airplane: :small_airplane: 

addresses (but doesn't yet close #40)

and https://github.com/screencloud/auth-next-frontend/issues/131

So to summarize few issues here:

-   We send messages with any targetOrigin
-   We receive messages and only verify the source

<a id="org3ac1ff3"></a>

# Proposed solution

The whitelist approach is the simplest one I could think of, but as Michael
raises, this won&rsquo;t work if customers are hosting their own apps. I don&rsquo;t think
it would be *too* difficult to just manually add customer domains as needed, as it&rsquo;s
doubtful this would grow into 4 digits or higher.

The valid domains list could come from somewhere else, e.g. `process.env` or
similar. In the function here is probably not the best place. It would be easy
to forget to change this and have accidental breakage.

I don&rsquo;t think this can come from the app `manifest.json` because the `start_url`
key in these are set to &ldquo;.&rdquo; This &ldquo;.&rdquo; will just grab the `index.html` in the
current folder, which itself has a `%PUBLIC_URL%` template string filled in by
`process.env.PUBLIC_URL`.

It&rsquo;s also not necessarily even used by the user agent.

> The start<sub>url</sub> member is purely advisory, and a user agent may ignore it or allow the user to alter it at install time or afterwards

<a id="org901dd59"></a>

## Message sending

For  the sending of messages, it seems that we would need to run
`this.target.postMessage(request, "*")` multiple times for each `send` event with the `*` replaced
with each potential source domain. Invalid domains would not receive this.

Otherwise, per my reading, browser extensions, etc. could receive this message.
Not sure if we&rsquo;re doing anything sensitive in these messages in particular.

<a id="orge8539cd"></a>

## Message receiving

Should be solved by checking `event.origin` against whitelist

<a id="orgca93dfa"></a>

# Other potential solutions:

1.  Split implementations so that there is a parentBridge and childBridge so that we

always know that the child&rsquo;s information is trustworthy, and the parent&rsquo;s is
not necessarily. As it stands with only one implementation for both, we cannot do this.

1.  Something something store it in the jwt. Not sure what this looks like given
    that you would still need a whitelist to check against the domain in the jwt.

# Update:

Proper domain list already exists in https://github.com/screencloud/auth-next-frontend/blob/master/src/appConfig.ts

### CODE REVIEW

**Comment by michael-martin** [comment]
The concept all looks good, but I think for apps, you can replace the whitelist with the values the app itself provides.

Every app includes a `viewerUrl` (i.e. the starting URL to load on the player) and optionally an `editorUrl` (i.e. the URL to load for editor apps to open the config page)

The player has that information already, so in theory it could check both values, strip back to just the domain (so no `/config` etc at the end of the value), and use that as part of the whitelist?

Then you're future-proofed against 3rd party devs and any potential changes to the apps domains we use too.

What do you think?

**Comment by vic08** [comment]
I like the idea of splitting into parentBridge and childBridge. Cause parent really doesn't need any whitelist check

File: src/PostMessageBridge.ts [hunk 1]
```diff
@@ -55,6 +74,7 @@ export class PostMessageBridge extends Bridge {
         targetWindow: Window | null = null,
         sourceWindow: Window = window,
         timeout: number = 1000,
+        verifyDomain: (options: IVerifyDomain) => boolean = defaultVerifyDomain,
     ) {
```
Comment (Macavirus) [inline src/PostMessageBridge.ts]: Not sure this function injection makes so much sense. What other kind of function would we expect to use other than a regex match against the whitelisted origins? It just means that we potentially need to redefine it / import it in every class that instantiates a PostmessageBridge.

**Comment by Macavirus** [comment]
> I like the idea of splitting into parentBridge and childBridge. Cause parent really doesn't need any whitelist check

Parent will still need to whitelist sending, otherwise it will broadcast messages that can be read by, e.g. any installed browser extension

**Comment by Macavirus** [comment]
> `viewerUrl`

I couldn't find any reference to `viewerUrl` in the apps repository. Has it been renamed maybe?

**Comment by vic08** [comment]
@Macavirus `viewerUrl` is used as src attribute for an iframe by the container

**Comment by vic08** [comment]
@Macavirus  I still don't get why does the parent need a whitelist? It gets the target url from the database, renders the iframe and sends the postmessage to that iframe's window

**Comment by luke** [comment]
<rant>It drives me crazy this repo is called app-core when it should be post message bridge or something like that.</rant>

All for adding the whitelist of domains or domain check code but we shouldn't be baking that list into this lib. Its mixing concerns. We should pass the list of domains or function to check when we use this lib. E.g. the list should be in auth.

**Comment by Macavirus** [comment]
https://github.com/screencloud/auth-next-frontend/pull/155 may fix this for auth (which is most important aspect to fix)

**Pull request closed without merging**